doc/src/manual/methods.md

```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+### Iterated dispatch
+
+In order to dispatch a multi-level parametric argument list,
+often it is best to separate each level of dispatch into separate functions.
+This may sound similar in approach to single-dispatch, but as we shall see below, it is still more flexible.
+
+For example, trying to dispatch on the element-type of an array will often run into ambiguous situations.
+Instead, commonly code will dispatch first on the container type,
+then recurse down to a more specific method based on eltype.
+In most cases, the algorithms lend themselves conveniently to this hierarchical approach,
+while in other cases, this rigor must be resolved manually.
+This dispatching branching can be observed, for example, in the logic to sum two matrices:
+
+```julia
+# First dispatch selects the map algorithm for element-wise summation.
++(a::Matrix, b::Matrix) = map(+, a, b)
+# Then dispatch handles each element and selects the appropriate
+# common element type for the computation.
++(a, b) = +(promote(a, b)...)
+# Once the elements have the same type, they can be added.
+# For example, via primitive operations exposed by the processor.
++(a::Float, b::Float) = Core.add(a, b)
+```
+
+### Trait-based dispatch
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
+We could do this by writing out a `Union` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
+it is possible to express type behaviors in the abstract
+in a way that is flexible to new additions, but which has no performance impact.
+
+A trait is implemented by defining a pure generic function which returns a
+singleton value from a particular set, as a computation on the types of its arguments.
+
+The example above glossed over the implementation details of [`map`](@ref) and [`promote`](@ref),
+since those both operate in terms of type traits.
+When iterating over a matrix, such as in the implementation of `map`,
+one important question is what order to use to traverse the data.
+When `AbstractArray` subtypes implement the `Base.IndexStype` trait,
+other functions such as `map` can dispatch on this information to pick
+the best algorithm (see [Abstract Array Interface](@ref man-interface-array)).
+This means that each subtypes does not need to implement a custom version of `map`,
+since the generic definitions + trait classes will enable the system to select the fastest version.
+
+```julia
+map(a::AbstractArray, b::AbstractArray) = map(Base.IndexStyle(a, b), a, b)
+map(::Base.IndexCartesian, a::AbstractArray, b::AbstractArray) = # generic implementation
+map(::Base.IndexLinear, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+```
+
+This trait-based mechanism is also present in the [`promote`](@ref) mechanism used by `+`.
+That mechanism uses the types to compute the optimal common type for computing the operation.
```

Comment:
_It uses the types of arguments to deduce the optimal common type for the result._?